rpc/service/src/service.rs

```diff
@@ -646,6 +654,91 @@ NOTE: This error usually indicates an RPC conversion error between the node and
+    ) -> RpcResult<GetPriorityFeeEstimateResponse> {
+        trace!("incoming GetPriorityFeeEstimateRequest request");
+
+        const CACHE_FOR: Duration = Duration::from_secs(1);
+
+        let relaxed_cache_resp = || GetPriorityFeeEstimateResponse {
+            fee_per_mass: FeePerMass::VirtualFeePerMass(VirtualFeePerMass {
+                max: self.estimated_fee_cache.max.load(Ordering::Relaxed),
+                median: self.estimated_fee_cache.median.load(Ordering::Relaxed),
+                min: self.estimated_fee_cache.min.load(Ordering::Relaxed),
+            }),
+            mempool_total_mass: self.mining_manager.snapshot().total_mass,
+        };
+        if self.estimated_fee_cache.expired_at.load(Ordering::Relaxed) > unix_now() {
+            debug!("Cache is not expired, return response from cache");
+            return Ok(relaxed_cache_resp());
+        }
+        if self
+            .estimated_fee_cache
+            .computing_in_progress
+            .compare_exchange_weak(false, true, Ordering::SeqCst, Ordering::Relaxed)
+            .is_err()
+        {
+            debug!("Computation is in progress, return response from cache");
+            return Ok(relaxed_cache_resp());
+        }
+
+        let compute = || async {
+            let script_public_key =
+                kaspa_txscript::pay_to_address_script(&Address::new(self.config.net.into(), Version::PubKey, &[0u8; 32]));
+            let miner_data: MinerData = MinerData::new(script_public_key, vec![]);
+            let session = self.consensus_manager.consensus().unguarded_session();
+            let kaspa_consensus_core::block::BlockTemplate {
+                block: kaspa_consensus_core::block::MutableBlock { transactions, .. },
+                calculated_fees,
+                ..
+            } = self.mining_manager.clone().get_block_template(&session, miner_data).await?;
+            // don't count coinbase tx
+            if transactions.len() < 2 {
+                debug!("Block template has no txs, return response from cache");
+                return Ok(relaxed_cache_resp());
+            }
+            let transactions = &transactions[1..]; // skip coinbase tx
+            let mut fees_and_masses = Vec::with_capacity(transactions.len());
+            for (mass, fee) in transactions.iter().map(Transaction::mass).zip(calculated_fees) {
+                fees_and_masses.push((fee, mass));
+            }
+            fees_and_masses
+                .sort_unstable_by(|(lhs_fee, lhs_mass), (rhs_fee, rhs_mass)| lhs_fee.mul(rhs_mass).cmp(&rhs_fee.mul(lhs_mass)));
+            let max = {
+                let (fee, mass) = fees_and_masses.last().unwrap();
+                *fee as f64 / *mass as f64
+            };
+            let min = {
+                let (fee, mass) = fees_and_masses.first().unwrap();
+                *fee as f64 / *mass as f64
+            };
+            let (div, rem) = fees_and_masses.len().div_rem(2);
```

Comment:
I think that in the even case we can simply take the first entry (so simply using `fees_and_masses[fees_and_masses.len()/2]`). It seems like averaging the two middle values is an overkill, and to me it looses the property of the median of being an actually used fee/mass ratio value (unlike the avg).